### PR TITLE
use ParamSpec to enable intellisense with distributed_trace decorators

### PR DESCRIPTION
**Pull request by kristapratico**
This PR is related to this [discussion](https://teams.microsoft.com/l/message/19:b97d98e6d22c41e0970a1150b484d935@thread.skype/1643319956805?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1643319956805&teamName=Azure%20SDK&channelName=Language%20-%20Python&createdTime=1643319956805) on Teams where we noticed that our `distributed_trace` decorators were blocking the intellisense type help. Fortunately there is a solution to this called typing.ParamSpec which forwards the parameter types of one callable to another callable and restores our type hinting!

ParamSpec: https://docs.python.org/3/library/typing.html#typing.ParamSpec

~This only works for py3.10+, but does enable intellisense as expected:~ This works for all supported Python versions.

VSCode/pylance:

image

Pycharm:

image

### CODE REVIEW

File: sdk/core/azure-core/azure/core/tracing/decorator.py [hunk 1]
```diff
@@ -25,39 +25,38 @@
...
 
-try:
-    from typing import TYPE_CHECKING
-except ImportError:
-    TYPE_CHECKING = False
+if sys.version_info >= (3, 10):
+    from typing import ParamSpec
+    P = ParamSpec("P")
+else:
+    P = ...
```
Comment (kristapratico) [inline sdk/core/azure-core/azure/core/tracing/decorator.py:40]: Prefer to do try/except, but it breaks intellisense that way.

**Comment by kristapratico** [comment]
We run mypy==0.782 in the CI and it erroneously calls out these errors in this PR:

```
azure/core/tracing/decorator_async.py:31: error: Module 'typing_extensions' has no attribute 'ParamSpec'
azure/core/tracing/decorator_async.py:41: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator_async.py:42: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator_async.py:49: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator_async.py:54: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator_async.py:66: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator.py:31: error: Module 'typing_extensions' has no attribute 'ParamSpec'
azure/core/tracing/decorator.py:41: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator.py:48: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator.py:53: error: The first argument to Callable must be a list of types or "..."
azure/core/tracing/decorator.py:65: error: The first argument to Callable must be a list of types or "..."
```

Running with the latest release (mypy==0.931) does not raise these errors. Options I see...

1) bump mypy version across the repo. Every package opted-in to mypy checks might start seeing some new errors which will need resolve: https://github.com/Azure/azure-sdk-for-python/blob/main/eng/tox/mypy_hard_failure_packages.py#L8

2) `type: ignore` these specific lines.

3) Add a mypy.ini to azure-core to exclude these files from mypy checks for now.

4) Is there a way to run a different version of mypy only on azure-core? Other options?

**Comment by kristapratico** [comment]
Output on mypy [opted-in packages](https://github.com/Azure/azure-sdk-for-python/blob/main/eng/tox/mypy_hard_failure_packages.py#L8) using mypy==0.931:

`azure-core` -
azure\core\settings.py:164: error: Incompatible return value type (got "Union[str, Type[AbstractSpan]]", expected "Optional[Type[AbstractSpan]]")

`azure-eventhub` -
azure\eventhub\_client_base.py:49: error: First argument to namedtuple() should be "_Address", not "Address"

`azure-keyvault-certificates` - 
azure\keyvault\certificates\_shared\_polling.py:50: error: Argument 1 of "wait" is incompatible with supertype "LROPoller"; supertype defines the argument type as "Optional[float]"
azure\keyvault\certificates\_shared\_polling.py:50: note: This violates the Liskov substitution principle
azure\keyvault\certificates\_shared\_polling.py:50: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides

`azure-keyvault-keys` -
azure\keyvault\keys\_shared\_polling.py:50: error: Argument 1 of "wait" is incompatible with supertype "LROPoller"; supertype defines the argument type as "Optional[float]"
azure\keyvault\keys\_shared\_polling.py:50: note: This violates the Liskov substitution principle
azure\keyvault\keys\_shared\_polling.py:50: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides

`azure-keyvault-secrets` -
azure\keyvault\secrets\_shared\_polling.py:50: error: Argument 1 of "wait" is incompatible with supertype "LROPoller"; supertype defines the argument type as "Optional[float]"
azure\keyvault\secrets\_shared\_polling.py:50: note: This violates the Liskov substitution principle
azure\keyvault\secrets\_shared\_polling.py:50: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides


`azure-servicebus` -
azure\servicebus\_servicebus_receiver.py:695: error: Argument 1 to "len" has incompatible type "Union[int, List[int]]"; expected "Sized"
azure\servicebus\_servicebus_receiver.py:705: error: Item "int" of "Union[int, List[int]]" has no attribute "__iter__" (not iterable)
azure\servicebus\aio\_servicebus_receiver_async.py:681: error: Argument 1 to "len" has incompatible type "Union[int, List[int]]"; expected "Sized"
azure\servicebus\aio\_servicebus_receiver_async.py:691: error: Item "int" of "Union[int, List[int]]" has no attribute "__iter__" (not iterable)

`azure-ai-metricsadvisor` -
azure\ai\metricsadvisor\_metrics_advisor_administration_client.py:642: error: syntax error in type comment "(*str, Any) -> None"
azure\ai\metricsadvisor\_metrics_advisor_administration_client.py:668: error: syntax error in type comment "(*str, Any) -> None"
azure\ai\metricsadvisor\_metrics_advisor_administration_client.py:694: error: syntax error in type comment "(*str, Any) -> None"
azure\ai\metricsadvisor\_metrics_advisor_administration_client.py:718: error: syntax error in type comment "(*str, Any) -> None"
azure\ai\metricsadvisor\_metrics_advisor_administration_client.py:1472: error: syntax error in type comment "(*str, Any) -> None"


`azure-identity`  - No changes needed
`azure-keyvault-administration` - No changes needed
`azure-ai-textanalytics`  - No changes needed
`azure-ai-formrecognizer`  - No changes needed
`azure-ai-translation-document`  - No changes needed
`azure-eventgrid`   - No changes needed
`azure-appconfiguration` - No changes needed
`azure-data-tables`  - No changes needed
`azure-mixedreality-remoterendering`  - No changes needed

**Review by iscai-msft — APPROVED** [review]

**Review by lmazuel — APPROVED** [review]

**Comment by xiaoyongzhu** [comment]
@kristapratico seems this PR introduces some import issue in databricks ML runtime (see this issue: https://github.com/linkedin/feathr/issues/154). Also seems like ParamSpec is added in 3.10 (https://docs.python.org/3/library/typing.html#typing.ParamSpec) so maybe we should also have a way to relax this limitation?

**Comment by xiaoyongzhu** [comment]
> @kristapratico seems this PR introduces some import issue in databricks ML runtime (see this issue: [linkedin/feathr#154](https://github.com/linkedin/feathr/issues/154)). Also seems like ParamSpec is added in 3.10 (https://docs.python.org/3/library/typing.html#typing.ParamSpec) so maybe we should also have a way to relax this limitation?

No worries, I think it's probably because of a lower typing extension in Databricks ML runtime.

**Comment by xiaoyongzhu** [comment]
And can be repoduced by this (in databricks ML 10.2):
```!pip install azure-core==1.23.1```
```from azure.core.tracing.decorator import distributed_trace```

and will throw out this error:
```

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<command-3599966980702568> in <module>
----> 1 from azure.core.tracing.decorator import distributed_trace

/databricks/python_shell/dbruntime/PythonPackageImportsInstrumentation/__init__.py in import_patch(name, globals, locals, fromlist, level)
    160             # Import the desired module. If you’re seeing this while debugging a failed import,
    161             # look at preceding stack frames for relevant error information.
--> 162             original_result = python_builtin_import(name, globals, locals, fromlist, level)
    163 
    164             is_root_import = thread_local._nest_level == 1

/databricks/python/lib/python3.8/site-packages/azure/core/tracing/decorator.py in <module>
     29 
     30 from typing import Callable, Any, TypeVar, overload
---> 31 from typing_extensions import ParamSpec
     32 from .common import change_context, get_function_and_class_name
     33 from ..settings import settings

ImportError: cannot import name 'ParamSpec' from 'typing_extensions' (/databricks/python/lib/python3.8/site-packages/typing_extensions.py)
```

**Comment by kristapratico** [comment]
@xiaoyongzhu moving this conversation to #23697

**Comment by glaubitz** [comment]
@kristapratico @lmazuel Is there a particular reason why you are depending on `typing_extensions >= 4.0.1`? Would version 3.10.0.0 work as well?

I am asking because I am trying to upgrade `azure-storage-blob` in SUSE SLE-15-SP1 to 12.13.1 which requires `azure-core` 1.23.1 which in turn introduced the dependency on `typing_extensions`.

We currently have `typing_extensions` 3.10.0.0 in SLE-15-SP1 and upgrading that to 4.0.1 would be rather difficult as that pulls in a number of new Python packages, so if we could stay with 3.10.0.0, it would make the upgrade for us much easier.

I did some tests with the Azure SDK and CLI with `typing_extensions` 3.10.0.0 and it seems to work but I am not 100% confident.

**Comment by kristapratico** [comment]
Hey @glaubitz, if I recall correctly `ParamSpec` was listed as an experimental feature in 3.10.0.0. I wouldn't expect any runtime errors, but I'm not sure about typing degradation due to the older version.

**Comment by kristapratico** [comment]
> @kristapratico Thanks for the answer! Do you know whether is a way to reliably test the compatibility?

@glaubitz here's what I did to test compatibility.

- created a virtual environment with Python 3.7 and installed the following versions:
    - azure-core==1.23.1
    - azure-storage-blob==12.13.1
    - typing-extensions==3.10.0.0
    - mypy==0.982

- created a simple sample with the azure-storage-blob BlobServiceClient where we pass in the wrong type for the parameter `analytics_logging`:

```python
from azure.storage.blob import BlobServiceClient
client = BlobServiceClient(account_url="x")

client.set_service_properties(analytics_logging=True)
```

- ran mypy on the above code and got an error: Argument "analytics_logging" to "set_service_properties" of "BlobServiceClient" has incompatible type "bool"; expected "Optional[BlobAnalyticsLogging]"

If you see the typing error, then typing is working correctly with the older version of typing-extensions (3.10.0.0) and using ParamSpec to forward the argument types to the client method. If you don't see the error (parameter gets typed as Any), then the typing experience is degraded with the older version. Note that I only tested this with Python 3.7 (although it appears to work on machine).

**Comment by glaubitz** [comment]
@kristapratico Wow, thanks a lot! This is incredibly helpful! I will perform these tests with the various Python versions.

**Comment by glaubitz** [comment]
Apologies for the late reply. I performed the check with `mypy` on Python3.6 and `typing_extensions == 3.10.0.0` and the type check fails as expected:

```
suse_gce@sle-15-jpag-test:~> mypy azure_blob_test.py 
azure_blob_test.py:4: error: Argument "analytics_logging" to "set_service_properties" of "BlobServiceClient" has incompatible type "bool"; expected "Optional[BlobAnalyticsLogging]"
Found 1 error in 1 file (checked 1 source file)
suse_gce@sle-15-jpag-test:~>
```

**Pull request merged**